File: solutions/_1443_minimum_time_to_collect_all_apples_in_a_tree/solution.py
from typing import List


class Solution:
    def minTime(self, n: int, edges: List[List[int]], hasApple: List[bool]) -> int:
        if n <= 1:
            return 0

        graph = {}
        for u, v in edges:
            if u not in graph:
                graph[u] = []
            if v not in graph:
                graph[v] = []
            graph[u].append(v)
            graph[v].append(u)

        def dfs(node, prev):
            res = 0
            for neighbor in graph[node]:
                if neighbor == prev:
                    continue
                res += dfs(neighbor, node)
            if (res > 0 or hasApple[node]) and node != 0:
                res += 2
            return res

        return max(0, dfs(0, -1))


# https://leetcode.com/problems/minimum-time-to-collect-all-apples-in-a-tree/

File: solutions/_1443_minimum_time_to_collect_all_apples_in_a_tree/solution.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {boolean[]} hasApple
 * @return {number}
 */
var minTime = function (n, edges, hasApple) {
  if (n <= 1) return 0;

  const graph = new Map();
  for (let i = 0; i < edges.length; i++) {
    const [u, v] = edges[i];
    if (!graph.has(u)) graph.set(u, []);
    if (!graph.has(v)) graph.set(v, []);
    graph.get(u).push(v);
    graph.get(v).push(u);
  }

  const dfs = (node, prev) => {
    let res = 0;
    for (const neighbor of graph.get(node)) {
      if (neighbor === prev) continue;
      res += dfs(neighbor, node);
    }
    if ((res > 0 || hasApple[node]) && node != 0) {
      res += 2;
    }
    return res;
  };

  return Math.max(0, dfs(0));
};

module.exports = { minTime };

/*
https://leetcode.com/problems/minimum-time-to-collect-all-apples-in-a-tree/
*/
